storage/src/backend/connection.rs

```diff
@@ -327,6 +377,93 @@ impl Connection {
             }
         }
 
+        let current_mirror = self.mirror_state.current.load();
+
+        if current_mirror.is_some() {
+            let mirror = current_mirror.as_ref().unwrap();
+            let data_cloned: Option<ReqBody<R>> = match data.as_ref() {
+                Some(ReqBody::Form(form)) => Some(ReqBody::Form(form.clone())),
+                Some(ReqBody::Buf(buf)) => Some(ReqBody::Buf(buf.clone())),
+                _ => None,
+            };
+
+            let mirror_host = Url::parse(&mirror.config.host).map_err(|e| {
+                ConnectionError::ErrorWithMsg(format!(
+                    "parse mirror host {} failed, {}",
+                    &mirror.config.host, e
+                ))
+            })?;
+
+            let mut current_url = Url::parse(url).map_err(|e| {
+                ConnectionError::ErrorWithMsg(format!("parse url {} failed, {}", url, e))
+            })?;
+            current_url.set_scheme(mirror_host.scheme()).map_err(|e| {
+                ConnectionError::ErrorWithMsg(format!("set mirror scheme failed, {:?}", e))
+            })?;
+            current_url.set_host(mirror_host.host_str()).map_err(|e| {
+                ConnectionError::ErrorWithMsg(format!("set mirror host failed, {:?}", e))
+            })?;
+            current_url.set_port(mirror_host.port()).map_err(|e| {
+                ConnectionError::ErrorWithMsg(format!("set mirror port failed, {:?}", e))
+            })?;
+
+            for (key, value) in mirror.config.headers.iter() {
+                headers.insert(
+                    HeaderName::from_str(key).unwrap(),
+                    HeaderValue::from_str(value).unwrap(),
+                );
+            }
+
+            let result = self.call_inner(
+                &mirror.client,
+                method.clone(),
+                current_url.to_string().as_str(),
+                &query,
+                data_cloned,
+                headers,
+                catch_status,
+                false,
+            );
+
+            match result {
+                Ok(resp) => {
+                    if resp.status() < StatusCode::INTERNAL_SERVER_ERROR {
+                        return Ok(resp);
+                    }
+                }
+                Err(err) => {
+                    warn!(
+                        "request mirror filed, mirror: {:?},  error: {:?}",
+                        mirror, err
+                    );
+
+                    let fail_limit = mirror.fail_limit.load(Ordering::Relaxed);
+                    if fail_limit > 0 {
+                        mirror.fail_limit.store(fail_limit - 1, Ordering::Relaxed);
+                    }
+
+                    if fail_limit == 1 && mirror.fail_limit.load(Ordering::Relaxed) <= 0 {
+                        error!("reach to fail limit, disable mirror: {:?}", mirror);
+                        mirror.status.store(false, Ordering::Relaxed);
```

Comment:
There still is race between load and store

Reply:
Already replaced it with `fetch_add()`.